bqskit/compiler/registry.py

```diff
@@ -0,0 +1,99 @@
+"""
+The _workflow_registry enables MachineModel or GateSet specific workflows to be
+registered for used in the `bqskit.compile` method.
+
+The _workflow_registry maps MachineModels a dictionary of Workflows which
+are indexed by optimization level. This object should not be accessed directly
+by the user, but instead through the `register_workflow` function.
+
+Example:
+    model_t = SpecificMachineModel(num_qudits, radixes)
+    workflow = [QuickPartitioner(3), NewFangledOptimization()]
+    register_workflow(model_t, workflow, level)
+    ...
+    new_circuit = compile(circuit, model_t, optimization_level=level)
+"""
+from __future__ import annotations
+
+import logging
+
+from bqskit.compiler.basepass import BasePass
+from bqskit.compiler.gateset import GateSet
+from bqskit.compiler.gateset import GateSetLike
+from bqskit.compiler.machine import MachineModel
+from bqskit.compiler.workflow import Workflow
+from bqskit.compiler.workflow import WorkflowLike
+from bqskit.ir.gate import Gate
+
+_logger = logging.getLogger(__name__)
+
+
+_workflow_registry: dict[MachineModel | GateSet, dict[int, Workflow]] = {}
```

Comment:
What's the design decision behind `MachineModel | GateSet` vs just `MachineModel`?

Reply:
`MachineModel` requires that you know beforehand how many qudits will be in some target, which is not something that is known when adding to the registry in namespace packages. Really what matters is just if the `GateSet` matches what is expected.

Reply:
This is true, but I think this feature's goal is entirely married to the MachineModel concept. By this, I mean an extension developer should be using this feature to automatically say: if you are compiling for this machine, this is a much better workflow than the off-the-shelf version

If you are sure your method is generic over a whole class of MachineModels, then you should have a `build_model` function exported with your package. At the end of this package, you can register the newly built model if its isn't registered yet.

Maybe I am missing something tho, lmk if you think there is a compelling use case for the GateSet. Either here or through Slack. My worry is with the potential conflicts/overlaps between GateSet and Model resolutions.